pdg/src/info.rs

```diff
@@ -96,52 +97,84 @@ fn get_last_desc(g: &mut Graph) -> HashMap<NodeId, NodeId> {
     desc_map
 }
 
-/// Finds the inverse of a [`Graph`], each [`Node`] mapping to a [`Vec`] of its children.
-fn collect_children(g: &Graph) -> HashMap<NodeId, Vec<NodeId>> {
-    let mut children = HashMap::<_, Vec<_>>::new();
+/// Finds the inverse of a [`Graph`], each [`Node`] mapping to a [`Vec`] of its immediate
+/// descendents and their relationship to the [`Node`] in terms of fields.
+fn collect_children(g: &Graph) -> HashMap<NodeId, Vec<(NodeId, Vec<Field>)>> {
+    let mut children = HashMap::<_, Vec<(NodeId, Vec<Field>)>>::new();
     for parent in g.nodes.indices() {
         children.entry(parent).or_default();
     }
-    for (parent, child) in g
+    for (parent, child, child_node) in g
         .nodes
         .iter_enumerated()
-        .filter_map(|(child, child_node)| Some((child_node.source?, child)))
+        .rev()
+        .filter_map(|(child, child_node)| Some((child_node.source?, child, child_node)))
     {
-        children.get_mut(&parent).unwrap().push(child);
+        if let NodeKind::Field(f) = child_node.kind {
+            let my_children =
+                children
+                    .remove(&child)
+                    .unwrap()
+                    .into_iter()
+                    .map(|(gchild, mut gchildf)| {
+                        (
+                            gchild,
+                            ({
+                                gchildf.push(f);
+                                gchildf
+                            }),
+                        )
+                    });
+            children.insert(child, Vec::new());
+            children.get_mut(&parent).unwrap().extend(my_children);
+        } else {
+            children
+                .get_mut(&parent)
+                .unwrap()
+                .push((child, Vec::<_>::new()));
+        }
     }
     children
 }
 
+fn prefix(fs1: &Vec<Field>, fs2: &Vec<Field>) -> bool {
+    let len1 = fs1.len();
+    let len2 = fs2.len();
+    for i in 0..min(len1, len2) {
+        if fs1[len1 - 1 - i] != fs2[len2 - 1 - i] {
+            return false;
+        }
+    }
+    true
+}
```

Comment:
Is this checking if one of `fs1` and `fs2` is a prefix of the other?

Reply:
Yes, and that's a job only efficient to do with prefix trees or some other data structure intended for it.

Reply:
This doesn't seem very inefficient?  It should be able to be compiled to just a `memcmp`.  And if performance is important, we can use `tinyvec` for the `Vec<Field>`s and the `memcmp` should be short and cache-resident, usually.  There shouldn't be very many `Field`s in a `Vec<Field>`, right?

Also, I can make this more idiomatic.

Reply:
The inefficient part isn't this, but having to run this for every single child. Though I suppose this is only inefficient in theory, because maybe there won't ever be any nodes with more than a few dozen immediate children.

Reply:
Ah, got it.  Well, let's save any optimizations until later when we know it's actually a bottleneck, or if a better trie solution is simpler.